portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,19 +14,71 @@
 
 package com.google.sps.servlets;
 
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.gson.Gson;
+import com.google.sps.data.Comment;
 import java.io.IOException;
+import java.text.SimpleDateFormat;
+import java.util.*;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
-@WebServlet("/data")
+@WebServlet("/comments")
 public class DataServlet extends HttpServlet {
+  private DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.ASCENDING);
+    PreparedQuery results = datastore.prepare(query);
+
+    int numLimit = Integer.parseInt(request.getParameter("limit"));
+    List<Entity> commentsEntity = results.asList(FetchOptions.Builder.withLimit(numLimit));
+    List<Comment> comments = new ArrayList<>();
+    for (Entity entity : commentsEntity) {
+      String name = (String) entity.getProperty("name");
+      String text = (String) entity.getProperty("text");
+      String timestamp = (String) entity.getProperty("timestamp");
+
+      comments.add(new Comment(name, text, timestamp));
+    }
+
+    response.setContentType("application/json;");
+    String json = new Gson().toJson(comments);
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Entity commentEntity = new Entity("Comment");
+    commentEntity.setProperty("name", request.getParameter("name-input"));
+    commentEntity.setProperty("text", request.getParameter("text-input"));
+
+    Date d = new Date();
+    SimpleDateFormat dateFormat = new SimpleDateFormat("HH:mm:ss MM/dd/yyyy");
+    commentEntity.setProperty("timestamp", dateFormat.format(d));
+
+    datastore.put(commentEntity);
+
+    response.sendRedirect("/index.html");
+  }
+
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+    String value = request.getParameter(name);
+    if (value == null) {
+      return defaultValue;
+    }
+    return value;
```

Comment:
Could inline this using ternary operators...

```suggestion
    return value == null ? defaultValue : value;
```

https://www.baeldung.com/java-ternary-operator